authorization/authorization.go

```diff
@@ -0,0 +1,161 @@
+package authorization
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"log"
+	"net/http"
+	"time"
+
+	jwtmiddleware "github.com/auth0/go-jwt-middleware"
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+type Handshake struct {
+	nonce      string
+	providerID string
+}
+
+// Outstanding handshakes.
+var handshakes map[string]Handshake
+
+func InitAuth() {
+	handshakes = make(map[string]Handshake)
+}
+
+// This should likely be replaced by middleware that validates the token's claims.
+func GetAuthMiddleware(signingKey []byte) *jwtmiddleware.JWTMiddleware {
+	return jwtmiddleware.New(jwtmiddleware.Options{
+		ValidationKeyGetter: func(token *jwt.Token) (interface{}, error) {
+			return signingKey, nil
+		},
+		SigningMethod: jwt.SigningMethodHS256,
+	})
+}
+
+type GetAuthChallengeResp struct {
+	Nonce string `json:"nonce"`
+}
+
+func GetAuthChallengeHandler(userIDString string, logger *log.Logger) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		providerIDs, present := r.URL.Query()[userIDString]
+		if !present {
+			w.WriteHeader(http.StatusBadRequest)
+			resp := AuthChallengeError{Message: "missing query value: providerID"}
+			json.NewEncoder(w).Encode(resp)
+			return
+		}
+
+		if len(providerIDs) < 1 {
+			w.WriteHeader(http.StatusBadRequest)
+			resp := AuthChallengeError{Message: "must specify a provider"}
+			json.NewEncoder(w).Encode(resp)
+			return
+		} else if len(providerIDs) > 1 {
+			w.WriteHeader(http.StatusBadRequest)
+			resp := AuthChallengeError{Message: "must specify only 1 provider"}
+			json.NewEncoder(w).Encode(resp)
+			return
+		}
+
+		providerID := providerIDs[0]
+
+		// Generate a nonce signed by the provider's public key
+		nonce := randString(8)
```

Comment:
We probably want to use 32 bytes, though I'm not sure how much that matters.
I would probably pull these out of the cryptographic random generator:

```
import (
   "crypto/rand"
)

nonce := make([]byte, 32)
_, err := rand.Read(nonce)
if err != nil {
   // ...
}
```

Reply:
Good point.